src/middleware/upload.ml

```diff
@@ -7,209 +7,118 @@
-let rec upload (request : Dream.request) =
-  let state = request.specific.upload in
-  match state.initial with
-  | true ->
-    state.initial <- false;
-
-    begin match Dream.header "Content-Type" request with
-    | Some content_type when check_content_type content_type ->
-
-      Lwt.async begin fun () ->
-        (* While waiting for the fields, Multipart_form_data will first receive
-           all file chunks. *)
-        let%lwt fields =
-          Multipart_form_data.parse
-            ~stream:(Lwt_stream.from (fun () -> Dream.read request))
-            ~content_type
-            ~callback:(chunk_callback state) in
-
-        (* All chunks have been received at this point. Now, process the form
-           fields. *)
-        let remaining_fields = ref fields in
-        let next_field () =
-          match !remaining_fields with
-          | [] -> None
-          | field::more ->
-            remaining_fields := more;
-            Some field
-        in
 
-        state.fields <- true;
-        state.field <- (fun () -> Lwt.return (next_field ()));
-
-        begin match state.event_listener with
-        | Some event_listener ->
-          begin match next_field () with
-          | Some field -> Lwt.wakeup_later event_listener (`Field field)
-          | None -> Lwt.wakeup_later event_listener `Done
-          end
-        | None ->
-          match state.chunk_listener with
-          | Some chunk_listener ->
-            Lwt.wakeup_later chunk_listener None
-          | None ->
-            ()
-        end;
-
-        Lwt.return_unit
-      end;
-
-      upload request
-
-    | _ ->
+let identify _ = object end
+
+let rec state (request : Dream.request) =
+  let stream = request.specific.upload.stream in
+  match%lwt Lwt_stream.peek stream with
+  | None -> let%lwt () = Lwt_stream.junk stream in Lwt.return `Done
+  | Some (uid, _header, _stream) when uid <> request.specific.upload.uid ->
+    log.debug (fun m -> m "Start to upload a new part.") ;
+    (* XXX(dinosaure): new part. *)
+    request.specific.upload.uid <- uid ;
+    request.specific.upload.nth <- 0 ;
+    request.specific.upload.name <- None ;
+    request.specific.upload.filename <- None ;
+    state request
+  | Some (_uid, header, _stream) ->
+    (* XXX(dinosaure): emit header. *)
+    let header = Multipart_form.Header.to_list header in
+    let n = request.specific.upload.nth in
+    if n < List.length header
+    then ( log.debug (fun m -> m "Show the field %d." n)
+         ; request.specific.upload.nth <- n + 1 
+         ; field_to_string request (List.nth header n) )
+    else
+      ( log.debug (fun m -> m "Start to upload contents.")
+      ; Lwt.return (`Part (request.specific.upload.name, request.specific.upload.filename)) )
+
+and upload (request : Dream.request) =
+  if request.specific.upload.uid = Lazy.force Dream.epsilon
+  then
+    let content_type =
+      let ( let* ) = Option.bind in
```

Comment:
I switched Dream to `let%lwt` to preserve users' stack traces as much as possible.

Reply:
Deleted any use of `let*` in my codebase.